### DISCUSSION TITLE
Analog Input

### DISCUSSION THREAD

**Discussion by Pandagineer** [bitcraze/discussions#264] [Q&A]
Hello. I would like to measure the voltage of an external sensor using an analog input. I have the prototype deck. (And also the flow deck and CrazyRadio, FYI.) I understand there are analog input channels A0 thru A4, and ground.

If I digress for the moment, I'd like to say that I love this tutorial: https://www.bitcraze.io/documentation/tutorials/getting-started-with-stem-drone-bundle/, because I more or less understand all the code. I ran it successfully.

Back to analog input, I found this post: https://forum.bitcraze.io/viewtopic.php?t=2908. It seems like I need the two *.h files, deck_analog.h and deck_constants.h. That's understood, but...

1) how do these two *.h files fit into the bigger picture? Do I need a third file as the "main program", which somehow calls them? Can I call them from the flow deck tutorial I mentioned above?
2) In the two *.h files, where do I specify the channel? i.e. let's say I use channel A2. Would I type "A2" somewhere?
3) Note that the flow deck tutorial is written in PY, whereas the two analog input files are in C. I'm generally confused on when to use PY vs C. My preference is python. Are there python versions of the deck_analog and deck_constants files?

Thanks for the help!

**Comment by krichardsson** [comment]
HI, and thanks for the kind words on the stem drone tutorial. We're happy you found it useful!

I'll start with your question 3: The simple answer is that you use python for code that run in the PC and C for code that runs in the Crazyflie. In this case you will have to write some code in C to make the Crazyflie read your analog input. I guess that you maybe also want to use the analog value in your PC and to make this work you will have to write some code in the Crazyflie (C) that "publishes" the data and some python code for the PC that reads the values from the Crazyflie an does something useful with it, like printing or graphing the value.

Question 1: One way to look at header files (in general) is that they describe other parts of the code, for instance the signature of a function or, as in this case some constants that can be used in a function. `deck_constants.h` describes (defines) the pins on the expansion bus, while `deck_analog.h` describes the functions used for analog pins. You need to include header files in .c files before you use a function or constant, to tell the compiler that they exist.
You can not use the .h files from the tutorial code, as this is python. 

Question 2: You include the header files at the top of a .c file and then use the functions and constants in your code. Something like this:
```c
#include "deck_constants.h"
#include "deck_analog.h""

float readMyAnalogPin() {
    return analogReadVoltage(DECK_GPIO_MISO);
}
```
Check the [Expansion port pinout](https://www.bitcraze.io/documentation/system/platform/cf2-expansiondecks/#expansion-port-pinout) to figure out which pin constant to use. In the example above `DECK_GPIO_MISO` is the same pin as `A3`

* You can read the [Getting started with development](https://www.bitcraze.io/documentation/tutorials/getting-started-with-development/) to learn how to build and flash C code to the Crazyflie.
* The easiest way for you is probably to write an [app in the Crazyflie](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/app_layer/)
* If you want to make your value available to a python scrip, the [logging frameworks is the easiest way](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/logparam/)

**Reply by Pandagineer** [reply]
Thank you for the great answers. Now would be a good time for me to explain what I'm trying to do.

I want to plug a thermocouple into my prototype deck, read the voltage, and then use that voltage to adjust pitch, or thrust, etc. Imagine flying towards (or away) from high temperature.

I have gone through your "Getting started with development" tutorial, and have also gone through other crazyflie-firmware projects. So I am familiar with looking at simple C code, compiling it, flashing the crazyflie, and watching the results. But before diving into your other two links (which are quite daunting), I would like to confirm the best path here, given my goal (described above):

- Should I go with C or Python, or a mixture of the two? (Although I prefer Python, I don't mind C. Also, is it unwise to mix the two languages?)
- Let's say the answer is C. I'm hoping that I can collect three C files: the two *.h files we discussed in original post, and a "main" program which calls them, measure the voltage, and commands the new flight path

You may be tempted to direct me to this project by the [University of Washington](https://www.bitcraze.io/2020/07/intro-to-autonomous-robotics-with-the-crazyflie/). I'm aware of this project and have even been in contact with some of the team members. But I am overwhelmed by the code, talk of API and ROS. I don't know where to start. This is why I'm approaching Bitcraze with some basic questions.

I would like to mention that I'm an engineer, but not a computer scientist or roboticist. I write and use Matlab code in my daily life to analyze data, integrate differential equations, manage files, etc. Simple stuff. But I rarely write functions, and almost never "include" multiple scripts. This is new ground for me, so I am trying to keep the program as simple as possible.

Thanks for your patience with these questions.

**Reply by tobbeanton** [reply]
I'll try to chip in a bit:
> Should I go with C or Python, or a mixture of the two? (Although I prefer Python, I don't mind C. Also, is it unwise to mix the two languages?)

As you prefer Python I would suggest to do the analog read-out in an app (C code). Save this in a logging variable that the PC can read over the radio using the python logging framework. The PC (python side) then runs the logic and commands the crazyflie what to do.

> Let's say the answer is C. I'm hoping that I can collect three C files: the two *.h files we discussed in original post, and a "main" program which calls them, measure the voltage, and commands the new flight path

Without writing it all for you :wink: I suggest copy the examples/app_hello_world example. First get that to run. Then modify it with what kricharsson suggested and save the voltage result into a variable. Make this variable available for the PC using the logging framework.

**Reply by Pandagineer** [reply]
Here is an update. I got "Hello World" to work. Next, because I want the Bitcraze tutorials to remain untouched, I copied "Hello World" to a new folder, and renamed it. More specifically, on the Desktop I created a new folder "Personal_Projects" and I copied *all* of the files from "..Crazyflie-firmware/examples/app_hello_world" to it, and then I renamed the app_hello_world folder to Measure_Voltage. 

But now when I try to run the Measure_Voltage project (which is just Hello World), it doesn't work. I know this because I changed the "Hello World" string slightly, but the Console is still showing the original "Hello World".

When I make the Measure_Voltage projects, I don't see any obvious errors. But perhaps I'm not looking closely enough.

Is there something about moving a project that breaks it? Please advise. Thanks!

**Reply by krichardsson** [reply]
Good that you got Hello world to work! 
We're trying to keep our discussions clean and prefer one question per thread. This thread is about analog input and is now sliding into out of tree builds. Would it be possible for you to create a new discussion for the app moving problem, and when that is solved we can continue to discuss analog inputs here?

**Reply by Pandagineer** [reply]
No problem. Thanks.

**Reply by Pandagineer** [reply]
I think I now have something which reads the voltage from a pin. I am storing this in variable x, which was declared as a float. I would like to check my voltage by printing x to the Console. But it seems that DEBUG_PRINT only accepts strings.

In C, how can I convert a float to string? This is easy in Python (use str), but in C I cannot figure it out. Please help.

**Reply by krichardsson** [reply]
This is a good question! Again, this is a new question, please start a new thread.

**Reply by Pandagineer** [reply]
The analog read code has been implemented. I am also outputting the read value to the Console. I have it on infinite loop, and every 2 sec the value of A3 is printed to the console. But (approximately) every 20 cycles of this, I get the message "ESTKALMAN: State Out of Bounds. Resetting". Should I be worried about this? Am I taking too many resources and so the Kalman filter cannot perform its job properly? I haven't actually tested the code while flying, so will it crash if the Kalman filter is reset?

Thanks!

**Reply by krichardsson** [reply]
No this is not a problem related to your implementation. If there is no positioning data sent to the kalman estimator, the noise from the IMU will make the estimate slowly drift away and when it is too unreasonable the supervisor will reset it. This is what you see.

**Reply by Pandagineer** [reply]
Ok. FYI, I ran observed the console again. Out of 165 echos, 6 of them were the Kalman reset (3.6%). Then I commented out all data collection in the code, and instead printed "hello world". Now out of 82 echos, zero were the Kalman reset (0%).

There certainly seems to be a correlation with what my code is doing. If the Kalman reset is not a problem, I'm ok. I just wanted to let you know.

**Reply by krichardsson** [reply]
Thanks!
If you shake the CF a bit you will probably get some more as well :-)

**Reply by Pandagineer** [reply]
Update. I am successfully measuring the voltage of an external 1.6 V battery between pins A3 and GND. (Yay!)

My next questions (let me know if I need to start new discussions): 

1. when I disconnect the external battery, the output sometimes goes to zero (roughly 10 mV), but then sometimes jumps to 3V. In practice, I will need the signal to stay at zero volts when I my sensor is reading nothing. (I don't mind averaging to filter noise, but if it jumps between zero and 3V, this would not work.)

2. When I apply -1.6V (i.e. I reverse the external battery), the Console stops echoing, and the M4 LED turns off. Will doing this damage the CF? If negative voltage is bad, how low is ok (i.e. can I give it -1 mV? -10 mV?) What is happening here?

For reference, my code is below. Thanks.

#include <string.h>
#include <stdint.h>
#include <stdbool.h>

#include "app.h"

#include "FreeRTOS.h"
#include "task.h"

#include "deck_constants.h"
#include "deck_analog.h"

float readMyAnalogPin() {
    return analogReadVoltage(DECK_GPIO_MISO);
}

#include "debug.h"

double y;

void appMain() {
  DEBUG_PRINT("Waiting for activation ...\n");
  
  while(1) {
    vTaskDelay(M2T(500));
    y = readMyAnalogPin();  
    DEBUG_PRINT("%f", y);
    DEBUG_PRINT("\n");
  
  }
}

**Reply by krichardsson** [reply]
Yes please start a new discussion :-)
